k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java

```diff
@@ -1023,6 +1028,32 @@ Upload any local messages that are marked as PENDING_UPLOAD (Drafts, Sent, Trash
 
     }
 
+    private void handleImapUidValidity(Account account, LocalFolder localFolder,
+                                       ImapFolder remoteImapFolder, String folder,
+                                       MessagingListener listener) throws MessagingException {
+        long previousUidValidity = localFolder.getLastUidValidity(account.getUuid(), remoteImapFolder.getName());
+        long currentUidValidity = remoteImapFolder.getCurrentUidValidity();
+
+        if (previousUidValidity != -1L && currentUidValidity != -1L && currentUidValidity != previousUidValidity) {
+
+            //Delete and reload all messages in the local folder
+            Set<String> localMessageUidSet = localFolder.getAllMessagesAndEffectiveDates().keySet();
```

Comment:
I wonder if it would be more appropriate to set the UIDs of all these messages to null, and leave deleting them to the normal syncing mechanism that relies on uids to do so. @cketti?